### PR TITLE
Improve docs about ARG in FROM

### PR DESCRIPTION
**Pull request by dnephin**
See https://github.com/docker/for-linux/issues/59 and https://github.com/moby/moby/pull/31352#issuecomment-315105319 and https://github.com/moby/moby/issues/34008

### CODE REVIEW

File: docs/reference/builder.md [hunk 1]
```diff
@@ -530,8 +530,10 @@ FROM extras:${CODE_VERSION}
 CMD  /code/run-extras
 ```
 
-To use the default value of an `ARG` declared before the first `FROM` use an
-`ARG` instruction without a value:
+An `ARG` declared before a `FROM` is outside of a build stage, so it
```
Comment (tonistiigi) [inline docs/reference/builder.md]: Maybe better to mention that every stage has individual args, then it is better to understand that "arg-in-from" is not an exception but just follows the rule.

How about this example:
```
ARG VERSION=latest
FROM busybox:$VERSION
ARG VERSION
RUN echo $VERSION > image_version
```

Comment (mstanleyjones) [inline docs/reference/builder.md]: Maybe something like "An `ARG` instructions is only relevant for the next `FROM` instruction that comes after it in the Dockerfile, and then using this example to illustrate that.

Comment (dnephin) [inline docs/reference/builder.md]: > Maybe something like "An ARG instructions is only relevant for the next FROM instruction that comes after it in the Dockerfile,

That's not exactly correct. There are two cases:
* any `ARG` before the first `FROM` can be used in any `FROM` line
* any `ARG` within a build stage (after a `FROM`) can be used in that build stage

I can add more details to the `ARG` reference section as well (this stuff is under `FROM`) , but I think this line needs to be updated as well something along the lines of what I have.

**Comment by dnephin** [comment]
I've updated to include more details, and added some more structure to the `ARG` section, PTAL

**Comment by dnephin** [comment]
Makes sense, I've swapped the examples.

**Comment by tonistiigi** [comment]
LGTM

**Pull request merged**